src/codegen/events/substrate.rs

```diff
@@ -32,68 +45,121 @@ impl EventEmitter for SubstrateEventEmitter<'_> {
-                let ty = arg.ty();
-
-                match ty {
-                    Type::String | Type::DynamicBytes => {
-                        let e = expression(
-                            &ast::Expression::Builtin(
-                                pt::Loc::Codegen,
-                                vec![Type::Bytes(32)],
-                                ast::Builtin::Keccak256,
-                                vec![arg.clone()],
-                            ),
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    Type::Struct(_) | Type::Array(..) => {
-                        // We should have an AbiEncodePackedPad
-                        let e = expression(
-                            &ast::Expression::Builtin(
-                                pt::Loc::Codegen,
-                                vec![Type::Bytes(32)],
-                                ast::Builtin::Keccak256,
-                                vec![ast::Expression::Builtin(
-                                    pt::Loc::Codegen,
-                                    vec![Type::DynamicBytes],
-                                    ast::Builtin::AbiEncodePacked,
-                                    vec![arg.clone()],
-                                )],
-                            ),
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    _ => {
-                        let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                        topics.push(e);
-                        topic_tys.push(ty);
-                    }
-                }
-            } else {
-                let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                data.push(e);
-                data_tys.push(arg.ty());
+        let loc = pt::Loc::Builtin;
+        let event = &self.ns.events[self.event_no];
+        let hash_len = || Box::new(Expression::NumberLiteral(loc, Type::Uint(32), 32.into()));
+
+        // Events that are not anonymous always have themselves as a topic.
+        // This is static and can be calculated at compile time.
+        if !event.anonymous {
+            let mut encoded =
+                format!("{}::{}", &self.ns.contracts[contract_no].name, &event.name).encode();
```

Comment:
Here you are using the contract name of the emitting contract, not where the event is defined. For example:

```solidity
contract A { event X(bool indexed foo); }

contract B { function f() public { emit A.X(true); } }
```
The topic you are using is `B.X`. This is not what the user might be expecting. However, if you want to use the name of the event definition, bare in mind that an event is not always defined in a contract:
```solidity
event X(bool indexed foo);

contract B { function f() public { emit X(true); } }
```
Whatever you decided needs documenting.

Reply:
Good catch!

Reply:
> Another thing to consider is, what happens when you emit an event in a free standing function:
> 
> ```solidity
> function foo() {
>     emit A(true);
> }
> ```

They are not a thing in `ink!`. They need to have a contract associated with them or we won't be compatible. But in that case we can actually just use the contract name and be fine.